Add Basis and Quadrature writing functionality to Volume Data

isaaclegred opened the pull request:
## Proposed changes

This PR modifies the volume data writing process to include writing the basis and quadrature used along each axis for each element.  So far this hasn't seemed to be a large performance drag but optimizations will be available if necessary to increase performance.  

### Types of changes:

- [ ] Bugfix
- [x ] New feature
- [ ] Refactor

### Component:

- [ x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

isaaclegred committed:
Add Basis and Quadrature writing functionality to Volume Data

isaaclegred commented:
Hey @nilsdeppe and @kidder I was hoping you could take a look at this, along with anyone else that might be interested in this functionality.

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -10,7 +10,8 @@
 
 #include "DataStructures/DataVector.hpp"
-
+// Could I forward Declare instead?
```
No. The reason is that you use `Basis` and `Quadrature` in the member variables of the definition of the `ElementVolumeData` struct, and therefore the definition of `Basis` and `Quadrature` are needed in order for the compiler to determine the size of the `ElementVolumeData` struct.

isaaclegred replied:
👍

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -61,3 +62,19 @@ struct ExtentsAndTensorVolumeData {
   std::vector<TensorComponent> tensor_components{};
 };
+
+// We need a datastructure to store the quadrature and the spectral basis as it
```
promote this to a doxygen comment like the other structs above

kidder commented on src/IO/CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ target_link_libraries(
   INTERFACE DataStructures
   INTERFACE ErrorHandling
   INTERFACE Utilities
```
alphabetize

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -3,6 +3,8 @@
 
 #include "IO/H5/VolumeData.hpp"
 
+#include <iostream>
```
just ostream

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -31,6 +35,7 @@ void append_element_extents_and_connectivity(
     const gsl::not_null<std::vector<int>*> total_connectivity,
     const gsl::not_null<int*> total_points_so_far, const size_t dim,
     const ExtentsAndTensorVolumeData& element) noexcept {
+  ;
```
remove

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -71,10 +76,29 @@ void append_element_name(const gsl::not_null<std::string*> grid_names,
       first_tensor_name.substr(0, first_tensor_name.find_last_of('/'));
   *grid_names += spatial_name + VolumeData::separator();
 }
+void append_element_quadratures(const gsl::not_null<std::string*> quadratures,
```
blank line before function definitions

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
       boost::make_transform_iterator(elements.front().tensor_components.end(),
                                      get_component_name));
+  // The dimension of the grid is the number of extents per element.
+  // Only written once per VolumeData file, as if two observation id's
```
I don't understand "as if two observation id's".  Please provide an example to make the part about dimension and extents clearer such as "if the extents are (4,5,9), the dimension of the grid is 3"

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
+  std::vector<int> total_connectivity;
+  std::string quadratures;
+  std::string bases;
+  // We need to keep track the total number of points inserted into the
```
have the comment say "keep a running count of the number of points so far to use as a global index for the connectivity"

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
+    std::string component_name = component_names[i];
+    std::vector<double> contiguous_tensor_data{};
+    for (auto& element : elements) {
+      if (i == 0) {  // True if first tensor component being accessed
```
use `UNLIKELY(i ==0)`

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
+
+    }  // for each element
+    // Write the data for the tensor component
+    if (h5::contains_dataset_or_group(observation_group.id(), "",
```
move this check to the beginning of the component loop

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
+  h5::write_data(observation_group.id(), total_connectivity,
+                 {total_connectivity.size()}, "connectivity");
+}
+// Write Volume Data stored in a vector of `ExtentsAndTensorVolumeData` to
```
add a blank line before the comment of the function definition

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -141,7 +165,108 @@ void VolumeData::write_volume_data(
+      boost::make_transform_iterator(elements.front().tensor_components.end(),
+                                     get_component_name));
   // The dimension of the grid is the number of extents per element.
   // Only written once per VolumeData file, as if two observation id's
```
again, see comment about this copied comment

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -200,6 +325,7 @@ void VolumeData::write_volume_data(
   std::vector<char> grid_names_as_chars(grid_names.begin(), grid_names.end());
   h5::write_data(observation_group.id(), grid_names_as_chars,
                  {grid_names_as_chars.size()}, "grid_names");
+  ;
```
this should be a blank line

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -310,5 +439,68 @@ size_t VolumeData::get_dimension() const noexcept {
+  const auto dim =
+      h5::read_value_attribute<size_t>(volume_data_group_.id(), "dimension");
+  const auto bases_per_element = static_cast<long>(dim);
+  std::cout << "working at VolumeData 445 inside get_bases"
```
remove the debugging statements

kidder commented on src/IO/H5/VolumeData.hpp:
```diff
@@ -117,7 +121,7 @@ class VolumeData : public h5::Object {
   /// Read the extents of all the grids stored in the file at the observation id
   /// `observation_id`
-  std::vector<std::vector<size_t>> get_extents(size_t observation_id) const
+  std::vector<std::vector<size_t> > get_extents(size_t observation_id) const
```
no space

kidder commented on src/IO/H5/VolumeData.hpp:
```diff
@@ -126,6 +130,14 @@ class VolumeData : public h5::Object {
   static char separator() noexcept { return ':'; }
 
+  /// Return the basis being used for each element along each axis
+  std::vector<std::vector<std::string> > get_bases(size_t observation_id) const
```
no extra space in template parameters (twice)

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -23,6 +24,7 @@
 #include "Parallel/Info.hpp"
 #include "Parallel/Invoke.hpp"
 #include "Utilities/Algorithm.hpp"
+#include "Utilities/MakeString.hpp"
```
this is not used

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -15,6 +15,7 @@
 #include "IO/H5/File.hpp"
 #include "IO/H5/VolumeData.hpp"
 #include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/Helpers.hpp"
```
check if this is used

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -142,7 +154,38 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+    // need them in their string form
+    auto target_bases = [&bases]() {
+      std::vector<std::vector<std::string>> local_target_bases;
+      for (size_t i = 0; i < 2; i++) {
```
why 2?  use an appropriate size or variable

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -142,7 +154,38 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+      std::vector<std::vector<std::string>> local_target_bases;
+      for (size_t i = 0; i < 2; i++) {
+        std::vector<std::string> axis_bases;
+        for (size_t j = 0; j < 3; j++) {
```
same with 3

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -142,7 +154,38 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+    std::sort(target_bases.begin(), target_bases.end(), std::less<>{});
+    auto target_quadratures = [&quadratures]() {
+      std::vector<std::vector<std::string>> local_target_quadratures;
+      for (size_t i = 0; i < 2; i++) {
```
again with 2 and 3

kidder reviewed (changes requested):
reviewed with @nilsdeppe

isaaclegred commented:
Sorry for force pushing, I wanted to fix the merge conflicts before they caused any more trouble down the line.

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -61,3 +62,19 @@ struct ExtentsAndTensorVolumeData {
+  ElementVolumeData() = default;
+  ElementVolumeData(std::vector<size_t> extents,
+                    std::vector<TensorComponent> components,
+                    std::vector<Spectral::Basis> basis,
```
basis -> basis_in to avoid shadowing... (same with quadrature)

isaaclegred commented:
@kidder and @nilsdeppe  I pushed a fixup whenever you get the chance to look.

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -270,8 +270,9 @@ void VolumeData::write_volume_data(
-  // The dimension of the grid is the number of extents per element.
-  // Only written once per VolumeData file, as if two observation id's
+  // The dimension of the grid is the number of extents per elements. I.e., if
+  // the extents are [3,5,7] then the dimension of the grid is 3.
```
don't use 3 in the `[3,5,7]` to avoid confusion with the dimension

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -168,8 +168,9 @@ void VolumeData::write_volume_data(
-  // Only written once per VolumeData file, as if two observation id's
+  // The dimension of the grid is the number of extents per elements. I.e., if
+  // the extents are [3,5,7] then the dimension of the grid is 3.
+  // Only written once VolumeData file.
```
once per

kidder reviewed (changes requested):
you can squash the minor things along with the fixup

kidder commented:
@isaaclegred is this your final version or were you planning to do more with this PR?  I ask because it is marked as `in progress`

isaaclegred commented:
@kidder I have removed the in progress label, I am happy for this to be merged whenever you and other reviewers think it is OK

kidder commented:
@isaaclegred this still has a shadowing error (see the gcc Travis builds), you can squash the fixup

nilsdeppe reviewed (dismissed):
Restarted travis because a whole bunch timed out in second stage

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -61,3 +61,22 @@ struct ExtentsAndTensorVolumeData {
+ */
+struct ElementVolumeData : ExtentsAndTensorVolumeData {
+  ElementVolumeData() = default;
+  ElementVolumeData(std::vector<size_t> exts,
```
exts -> extents

isaaclegred replied:
@kidder should I make this `in_extents`?  I feel like the statement `extents(std::move(extents))` seems a bit strange, but if it's the preferred way of handling this I can do it.

kidder replied:
sure, extents_in is fine (to be consistent with the other args)

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -61,3 +61,22 @@ struct ExtentsAndTensorVolumeData {
+      : ExtentsAndTensorVolumeData(std::move(exts), std::move(components)),
+        basis(std::move(basis_in)),
+        quadrature(std::move(quadrature_in)) {}
+  void pup(PUP::er& p) noexcept;  // NOLINT
```
blank line before pup

kidder commented on src/IO/CMakeLists.txt:
```diff
@@ -31,6 +31,7 @@ target_link_libraries(
   Hdf5
   Options
   Utilities
```
alphabetical

kidder commented on src/IO/H5/Python/CMakeLists.txt:
```diff
@@ -19,6 +19,7 @@ spectre_python_link_libraries(
   PRIVATE
   IO
   DataStructures
+  Informer
```
alphabetize

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+#pragma once
```
SpectralIO -> SpectralIo

kidder replied:
blank line before pragma

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+#include <boost/iterator/transform_iterator.hpp>
+#include <hdf5.h>
+#include <memory>
+#include <ostream>
```
check the includes

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+#include "IO/H5/VolumeData.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/MakeString.hpp"
+namespace h5 {
```
blank line before namespace

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+/// We maintain a list of bases and quadratures which are compatible
+/// with IO, this allows for efficient storage of Spectral::Basis in
+/// volume data files
+const std::vector<Spectral::Basis> IO_bases{Spectral::Basis::Chebyshev,
```
turn this into the function allowed_bases, put into namespace h5_detail and move implementation into cpp file

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+                                            Spectral::Basis::Legendre,
+                                            Spectral::Basis::FiniteDifference};
+/// A list of the names of the Bases, for use in a dictionary
+const std::vector<std::string> IO_basis_names = []() {
```
same for this or better try to use get_output

kidder replied:
noexcept

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+/// We maintain a list of quadratures which are compatible
+/// with IO, this allows for efficient storage of Spectral::Quadrature in
+/// volume data files
+const std::vector<Spectral::Quadrature> IO_quadratures{
```
same requests as with the bases

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+}();
+/// Write a dictionary as an attribute to the volume file, can be used
+/// to decode integer sequence as values[i] represents the string
+/// value encoded which was enconded with integer i
```
enconded -> encoded

kidder commented on src/IO/H5/SpectralIO.hpp:
```diff
@@ -0,0 +1,77 @@
+}
+/// A dictionary `dict_name` is used to decode the integer vector `decodable`
+/// into an vector of strings.  The `dict_name` should correspond to
+/// a dictionary written with `h5::write_dictonary`
```
write_dictionary

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -203,11 +225,22 @@ void VolumeData::write_volume_data(
+  h5::write_data(observation_group.id(), quadratures, {quadratures.size()},
+                 "quadratures");
+  // Write the coded basis, along with the dictionary
+  h5::write_dictionary("Basis dictionary", h5::IO_basis_names,
```
change these to the new implementation

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -203,11 +225,22 @@ void VolumeData::write_volume_data(
 }
 
+
+
```
one blank line

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -74,10 +79,27 @@ void append_element_name(const gsl::not_null<std::string*> grid_names,
+  for (size_t axis = 0; axis < dim; axis++) {
+    quadratures->push_back(static_cast<int>(element.quadrature[axis]));
+  }
+}
```
@nilsdeppe will add an `alg::transform` which can be used for the appends

kidder replied:
```
std::transform(qudrature.begin(), quadrature.end(), std::back_inserter(output.end()), [](const ENUM_TYPE t) { return static_cast<int>(t); });
```

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -231,14 +264,18 @@ std::vector<std::string> VolumeData::list_tensor_components(
-              "total_extents");
-  std::remove(tensor_components.begin(), tensor_components.end(), "grid_names");
+  auto remove_data_name = [&tensor_components](const std::string& data_name) {
+    return static_cast<void>(std::remove(tensor_components.begin(),
```
switch to `alg::remove`

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -321,7 +358,7 @@ std::pair<size_t, size_t> offset_and_length_for_grid(
     const size_t element_data_offset = std::accumulate(
         all_extents.begin(), all_extents.begin() + element_index,
-        0_st, [](size_t offset, std::vector<size_t> extents) noexcept {
+        0_st, [](size_t offset, const std::vector<size_t>& extents) noexcept {
```
const offset

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -334,5 +371,48 @@ size_t VolumeData::get_dimension() const noexcept {
 }
 
+std::vector<std::vector<std::string>> VolumeData::get_bases(
+    size_t observation_id) const noexcept {
```
const

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -334,5 +371,48 @@ size_t VolumeData::get_dimension() const noexcept {
+  return element_bases;
+}
+std::vector<std::vector<std::string>> VolumeData::get_quadratures(
+    size_t observation_id) const noexcept {
```
const

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -83,11 +87,18 @@ struct ContributeVolumeData {
-                    ExtentsAndTensorVolumeData(
-                        std::move(extents), std::move(received_tensor_data)));
+                .emplace(array_component_id,
+                         ElementVolumeData(std::move(extents),
```
instead of moving, pass the begin and end of extents, basis, quadrature

kidder commented on support/Environments/wheeler_clang.sh:
```diff
@@ -67,4 +67,5 @@ spectre_run_cmake() {
           -D CMAKE_CXX_LINK_FLAGS=-lstdc++ \
           "$@" \
           $SPECTRE_HOME
+
```
?

kidder commented on tests/Unit/IO/Importers/CMakeLists.txt:
```diff
@@ -8,7 +8,6 @@ set(LIBRARY_SOURCES
   Test_Tags.cpp
   Test_VolumeDataReaderActions.cpp
   )
-
```
?

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+    auto read_bases = volume_file.get_bases(observation_id);
+    std::sort(read_bases.begin(), read_bases.end(), std::less<>{});
+    auto read_quadratures = volume_file.get_quadratures(observation_id);
+    std::sort(read_quadratures.begin(), read_quadratures.end(), std::less<>{});
```
@nilsdeppe will add alg::sort

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+    // clang-tidy consider pre-allocating the vector capacity
+    auto target_bases = [&bases]() {
+      std::vector<std::vector<std::string>> local_target_bases;
+      for (auto element_bases : bases) {
```
const auto&

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+    // We need non-const bases and quadratures in order to sort them, and we
+    // need them in their string form,
+    // clang-tidy consider pre-allocating the vector capacity
+    auto target_bases = [&bases]() {
```
noexcept

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+      std::vector<std::vector<std::string>> local_target_bases;
+      for (auto element_bases : bases) {
+        std::vector<std::string> axis_bases;
+        for (auto axis_basis : element_bases) {
```
const auto&

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+      for (auto element_bases : bases) {
+        std::vector<std::string> axis_bases;
+        for (auto axis_basis : element_bases) {
+          axis_bases.emplace_back(MakeString{} << axis_basis);  // NOLINT
```
// NOLINTNEXTLINE(CLANG_TIDY_ERROR) on line above and remove // NOLINT

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -152,7 +164,40 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+      return local_target_bases;
+    }();
+    std::sort(target_bases.begin(), target_bases.end(), std::less<>{});
+    auto target_quadratures = [&quadratures]() {
```
same comments as for bases

kidder replied:
also try to combine the lambdas by declaring once with an argument and using it twice

isaaclegred replied:
@kidder or @nilsdeppe is there a standard procedure for making the one lambda work for both the bases and quadratures, doesn't seem like templates and lambdas work well together; someone on the internet suggested  `decltype`?

nilsdeppe replied:
You can take the argument as `const auto& basis_or_quadrature`. The `auto` is the way to get templates in lambda expressions

isaaclegred replied:
Great, thanks!

isaaclegred commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+
+/// A dictionary `dict_name` is used to decode the integer vector `decodable`
+/// into an vector of strings.  The `dict_name` should correspond to
+/// a dictionary written with `h5::write_dictionary`
```
`h5:: -> h5_detail::`

isaaclegred commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -265,8 +261,8 @@ std::vector<std::string> VolumeData::list_tensor_components(
   auto remove_data_name = [&tensor_components](const std::string& data_name) {
-    return static_cast<void>(std::remove(tensor_components.begin(),
-                                         tensor_components.end(), data_name));
+    // NOLINTNEXTLINE bugpront-unused-return-value
```
`bugpront` -> `bugprone`

isaaclegred commented:
@kidder and @nils, sorry I didn't realize squashing all my fixups into one fixup and force pushing it would destroy the history before any of the commits happened, I have learned my lesson.  If you want to take a look at the changes that would be great, also now this PR will be broken because of @isha1810's first PR getting merged I think, but I have another branch that has all the correct commit orders so maybe I can replace the PR with that branch soon?

isaaclegred commented:
@kidder and @nilsdeppe can you look at this so I can rebase it?

kidder commented on src/DataStructures/Tensor/TensorData.hpp:
```diff
@@ -10,7 +10,7 @@
 
 #include "DataStructures/DataVector.hpp"
-
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
```
blank line after includes

kidder commented on src/IO/Connectivity.cpp:
```diff
@@ -7,7 +7,6 @@
 
 #include "DataStructures/Index.hpp"  // IWYU pragma: keep
 #include "ErrorHandling/Error.hpp"
-
```
keep this blank line

isaaclegred replied:
I think something is weird here, I didn't change anything but I don't see a removed line on the diff

kidder commented on tests/Unit/Visualization/Python/Test_GenerateXdmf.py:
```diff
@@ -6,12 +6,92 @@
+    def setUp(self):
+        # We create a vol file using the python wrapper.
+        self.file_name = os.path.join(spectre_informer.unit_test_path(),
+                                      "Visualization/Python/VolTestData.h5")
```
is this filename unique to this test?

isaaclegred replied:
At the point this change will come in the commit order it will be, but this is just the same name as the file the that the test used before so it can be changed.

nilsdeppe replied:
Okay, it would be good to name it something include the name of the test. that way it will be unique.

Related: I remember we discussed at a call add a `unit_test_build_path` and ` unit_test_src_path` to avoid writing into the build directory. Where did we decide to add that and who was going to do it?

kidder commented on src/DataStructures/Python/CMakeLists.txt:
```diff
@@ -10,7 +10,7 @@ spectre_python_add_module(
   DataVector.cpp
   Matrix.cpp
-  TensorData.cpp
+  ${CMAKE_SOURCE_DIR}/src/DataStructures/Tensor/Python/TensorData.cpp
```
I'm puzzled as to why this change is needed

isaaclegred replied:
@kidder sorry I think another fixup commit got squashed into this one, but now this commit is out of date, I can't really fix it without doing an interactive rebase and incorporating the changes made to the previous pull requests.

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include <string>
+#include <vector>
+
+#include "ErrorHandling/Assert.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include <vector>
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/H5/AccessType.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Helpers.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Helpers.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "OpenGroup.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include "IO/H5/Helpers.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "OpenGroup.hpp"
+#include "Utilities/MakeString.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+
+#pragma once
+
+#include <hdf5.h>
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <hdf5.h>
+#include <ostream>
```
remove

kidder commented on src/IO/H5/SpectralIo.hpp:
```diff
@@ -0,0 +1,44 @@
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "OpenGroup.hpp"
+#include "Utilities/MakeString.hpp"
+
```
add forward declaration of OpenGroup

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Helpers.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "OpenGroup.hpp"
```
IO/H5/OpenGroup.hpp

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+#include "SpectralIo.hpp"
+
+#include <hdf5.h>
```
remove

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+#include "SpectralIo.hpp"
+
+#include <hdf5.h>
+#include <ostream>
```
remove

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+#include <string>
+#include <vector>
+
+#include "ErrorHandling/Assert.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+#include <vector>
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/H5/AccessType.hpp"
```
remove

kidder commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+#include "IO/H5/Helpers.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "OpenGroup.hpp"
+#include "Utilities/MakeString.hpp"
```
remove

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -199,8 +183,15 @@ void VolumeData::write_volume_data(
+        // append element quadraature
+        alg::transform(
+            element.quadrature, std::back_inserter(quadratures),
+            [](const Spectral::Quadrature t) { return static_cast<int>(t); });
```
noexcept

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -199,8 +183,15 @@ void VolumeData::write_volume_data(
+        // append element basis
+        alg::transform(
+            element.basis, std::back_inserter(bases),
+            [](const Spectral::Basis t) { return static_cast<int>(t); });
```
noexcept

kidder commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -165,37 +165,29 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
-    }();
-    std::sort(target_bases.begin(), target_bases.end(), std::less<>{});
+    const auto& stringify =
+        [](const auto& bases_or_quadratures) {
```
noexcept

kidder reviewed (commented):
okay, @nilsdeppe please look at, I'm fine with the current state

nilsdeppe commented on tests/Unit/IO/Test_VolumeData.cpp:
```diff
@@ -170,24 +170,22 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
+      for (const auto& element_data : bases_or_quadratures) {
+        std::vector<std::string> target_axis_data;
+        for (const auto& axis_datum : element_data) {
+          // NOLINTNEXTLINE clang-tidy consider pre-allocating
```
Can't this be addressed by getting the number of elements from`element_data`? E.g.
```cpp
target_axis_data.reserve(bases_or_quadratures.size() + 1);
```

nilsdeppe commented on src/IO/H5/SpectralIo.cpp:
```diff
@@ -0,0 +1,52 @@
+
+namespace h5_detail {
+
+const std::vector<Spectral::Basis> allowed_bases() noexcept {
```
sorry what is `const` here? A `const` return type doesn't change anything from just return by value.

Related: can this be made into a `std::array`? I'm not sure, but it would eliminate an allocation if it could be

isaaclegred replied:
Sorry the `const` was left over from when this was a variable and not a function.  It can be made into a `std::array` I'm pretty sure so I will change it.

nilsdeppe commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp:
```diff
@@ -256,7 +254,7 @@ struct ComplicatedSystem {
       check_component("Error(Tensor2)_yy", TensorVar2{}, 1, 1);
     }
-
+    // NOLINTNEXTLINE
```
If these do end up needed, can you `NOLINTNEXTLINE(error...)` to only NOLINT the specific errors? Also a few lines above

nilsdeppe reviewed (commented):
Please go ahead and squash and rebase, including anything addressing these last few minor comments. Thanks! :)

nilsdeppe commented:
you'll need to rebase for the checks to run in the PR, but they might still run on your fork

isaaclegred commented:
> you'll need to rebase for the checks to run in the PR, but they might still run on your fork

Thanks for the heads up!  Seems like on the fork it runs clang tidy on all of the files whereas in the PR it only runs on changed files, is this the expected behavior?

nilsdeppe commented:
That depends on the last time you rebased. It shouldn't be running on effectively all files, but if the branch is sufficiently old then you don't have the fix

isaaclegred commented:
@kidder and @nilsdeppe, I've rebased and updated the pull request, would you prefer to keep this as 2 commits or squash it into 1 commit?

isaaclegred commented:
@nilsdeppe @kidder  do you have any more requests/ want to group review again?

kidder commented on src/IO/CMakeLists.txt:
```diff
@@ -30,6 +30,8 @@ target_link_libraries(
   Domain
   DomainStructure
+  Spectral
+
```
no blank-line

kidder commented on src/DataStructures/Python/Tensor/TensorData.cpp:
```diff
@@ -45,5 +58,15 @@ void bind_tensordata(py::module& m) {  // NOLINT
       });
+  py::class_<ElementVolumeData, ExtentsAndTensorVolumeData>(m,
+                                                            "ElementVolumeData")
+
```
remove this blank line

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -10,20 +10,25 @@
 #include "Utilities/Literals.hpp"
+#include "Utilities/MakeString.hpp"
 #include "Utilities/Numeric.hpp"
-
```
keep this blank line

kidder commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -144,9 +150,9 @@ void VolumeData::write_volume_data(
-
-  // The dimension of the grid is the number of extents per element.
-  // Only written once per VolumeData file, as if two observation id's
+  // The dimension of the grid is the number of extents per elements. I.e., if
```
`per element`

isaaclegred commented:
@kidder I made the changes, do you mind if I squash and rebase on develop to fix the conflict?

kidder commented:
yes, you can squash

nilsdeppe commented on src/Executables/ExportCoordinates/ExportCoordinates.hpp:
```diff
@@ -108,8 +108,9 @@ struct ExportCoordinates {
-        std::move(components), mesh.extents());
+        std::move(components), mesh.extents(), mesh.basis(), mesh.quadrature());
     return std::forward_as_tuple(std::move(box));
+
```
no blank line

isaaclegred commented:
@nilsdeppe and @kidder there was a rather weird merge conflict that appeared inside of `VolumeActions.hpp` but I think everything should be up-to-date now I think this is ready for further review whenever you get the chance!

nilsdeppe reviewed (commented):
`SpectraIo.hpp` will need to be listed in the CMakeLists.txt. @isaaclegred I can take this over in a way that makes the commit give you credit. I think it's really just a rebase that's left, so you definitely need to get all the credit :) Is that okay?

It also looks like this is failing to build. Again, I'm happy to finish up the last tiny bit and make sure GitHub gives you credit :)

isaaclegred commented:
> `SpectraIo.hpp` will need to be listed in the CMakeLists.txt. @isaaclegred I can take this over in a way that makes the commit give you credit. I think it's really just a rebase that's left, so you definitely need to get all the credit :) Is that okay?
> 
> It also looks like this is failing to build. Again, I'm happy to finish up the last tiny bit and make sure GitHub gives you credit :)

Sounds good to me!  Let me know if you come across anything thats unclear!  (I fixed the space issue and added `SpectralIo.hpp` in a forced push last night, sorry if that was confusing)

nilsdeppe commented:
@isaaclegred this looks great. If you don't mind, could you squash it so we can merge it?

isaaclegred commented:
> @isaaclegred this looks great. If you don't mind, could you squash it so we can merge it?

Squashed!

nilsdeppe reviewed (approved):
Awesome! Thank you for doing this. Really nice work! :D

isaaclegred commented:
Thanks for all your help!
